### PR TITLE
on-disk async indexing

### PR DESCRIPTION
**Pull request by asdine**
### What's being changed:

This adds a complete rework of async indexing, including:
- a new generic `queue` package
- a new vector index queue that uses that package.

#### The queue package

The package contains two main types:
- Queue interface / DiskQueue implementation
- Scheduler

The DiskQueue persists enqueued tasks to disk in chunks. Each chunk is 100 MB in size.
Chunks contain a header that describes the format of the chunk and the number of records.

The Scheduler manages the processing of the queues. It spawns N workers upon startup (N = number of cores - 1).
Each queue registers itself to the Scheduler. The Scheduler then goes through each queue periodically and processes them.
Queue processing:
The Scheduler requests a batch from the queue, then partitions it by task key (typically the DocID).
Each partition is then sent to a dedicated worker for execution.
Then Scheduler then moves on to the next queue and repeats the process.

#### The Vector Index Queue

The VectorIndexQueue wraps the DiskQueue and deals with async indexing.
It defines how the insert / delete operations are encoded / decoded.
It also deals with compression by automatically pausing the queue and triggering compression when the index reaches a certain size.

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [x] Chaos pipeline run or not necessary. Link to pipeline: https://github.com/weaviate/weaviate-chaos-engineering/actions/runs/12140480342
- [x] e2e test suite: https://github.com/weaviate/weaviate-e2e-tests/actions/runs/12140984201
- [x] All new code is covered by tests where it is reasonable.
- [x] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/queue/queue.go [hunk 1]
```diff
@@ -0,0 +1,747 @@
...
+
+	return nil
+}
+
+func (q *DiskQueue) ID() string {
+	return q.id
+}
+
+func (q *DiskQueue) Push(record []byte) error {
+	if q.closed.Load() {
```
Comment (jeroiraz) [inline adapters/repos/db/queue/queue.go:194]: I think a lock may be preferred instead, otherwise the queue could be closed right after this validation passed

File: adapters/repos/db/queue/queue.go [hunk 2]
```diff
@@ -0,0 +1,747 @@
...
+		err := q.promoteChunkNoLock()
+		if err != nil {
+			return err
+		}
+	}
+
+	if q.f == nil {
+		// create or open partial chunk
+		path := filepath.Join(q.dir, partialChunkFile)
+		f, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY, 0644)
```
Comment (jeroiraz) [inline adapters/repos/db/queue/queue.go:243]: it may be convenient to also add `os. O_APPEND`

Comment (jeroiraz) [reply adapters/repos/db/queue/queue.go:243]: ah just saw below that it's not treated as append-only file

Comment (asdine) [reply adapters/repos/db/queue/queue.go:243]: Exactly, I started with O_APPEND, but I need to update the header before promoting the chunk so it doesn't work.

File: adapters/repos/db/queue/queue.go (hunk 1 above)
Comment (asdine) [reply adapters/repos/db/queue/queue.go:194]: That's a good point

**Comment by rlmanrique** [comment]
The E2E tests are failing. At least one cause has been identified, sometimes the executions of multiple test in a row may leave the system hanging. It looks like a lock problem.

E2E failing job: https://github.com/weaviate/weaviate-e2e-tests/actions/runs/11898121190

File: usecases/monitoring/prometheus.go [hunk 3]
```diff
@@ -451,46 +445,36 @@ func newPrometheusMetrics() *PrometheusMetrics {
 			Help: "Time in ms for a bucket operation to complete",
 		}, []string{"strategy", "class_name", "shard_name", "path", "operation"}),
 
+		// Queue metrics
+		QueueSize: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "queue_size",
```
[the remark this answers is not in the document]
Comment (asdine) [reply usecases/monitoring/prometheus.go:450]: That was on purpose actually, the queue package is completely generic and can be used for anything, so `vector_` can't work if we decide to use it for something else. WDYT?

File: adapters/repos/db/queue/queue.go [hunk 4]
```diff
@@ -0,0 +1,1046 @@
...
+
+	q.m.Lock()
+	if q.closed {
+		q.m.Unlock()
+		return errors.New("queue already closed")
+	}
+	q.closed = true
+	q.m.Unlock()
+
+	q.scheduler.UnregisterQueue(q.id)
```
Comment (abdelr) [inline adapters/repos/db/queue/queue.go:182]: I am trying to understand why we do the q.m.Unlock() just to unregister the queue and then lock again. I think it is only because the UnregisterQueue uses the Pause which in turn locks too, leading to a deadlock if it is already locked. I think in this path then we have three times a lock on the same mutex. Should be more efficient to keep it locked through the complete block, right? I do not think this is a hot path and I could be totally wrong so feel free to ignore. I have seen in some places that we do a nonblocking version of the function to avoid this, in this case we could have some PauseNoLock method that pauses the queue without explicitly locking since it assumes it is already locked. I feel like this could also be more error pruned but more efficient. Could you tell me what are your thoughts on this so I can try to be consistent later?

**Review by abdelr — DISMISSED** [review]

File: adapters/repos/db/queue/queue.go (hunk 4 above)
Comment (asdine) [reply adapters/repos/db/queue/queue.go:182]: This function is `Close`, so it's usually called only during the node shutdown, or when we drop a shard. So basically it's not a hot path, as long as it's functionally correct and easy to read (I hope), anything is fine.
Other important point: it's usually called only once, not concurrently.

(Then why lock if it's gonna be called once? It's to prevent Push if the queue is closed)

Then about locking: We first lock the `q.closed` variable so we can prevent concurrent access to the rest of the function. Only one goroutine can skip the if statement l.175, all the other ones return l.177.
So we know that we can unregister only once. (I could have used a `sync.Once` for example, it would have worked the same)

> I think in this path then we have three times a lock on the same mutex.

Not exactly: `UnregisterQueue` is not a method of the queue, it's a method of the `Scheduler`.

Because it's a completely different entity (there can be N queues, but only 1 scheduler), we cannot lock while we are asking the scheduler to deregister the queue: the scheduler could be calling the queue at the same time (for chunk deletion for example), which would lead to a deadlock.

Then, only when we know the queue is unregistered from the scheduler (i.e. the scheduler is guaranteed NOT to read anything from the queue) we can close the files.

Basically, even if it's one single function, it's 3 different steps:
- setting the `q.closed` to true, but only once, filtering out any other call to Close
- unregistering the queue: can take a while if the scheduler is busy
- closing the resources

File: adapters/repos/db/queue/scheduler.go [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+	opts.Logger = opts.Logger.WithField("component", "queue-scheduler")
+
+	if opts.Workers <= 0 {
+		opts.Workers = max(1, runtime.GOMAXPROCS(0)-1)
+	}
+
+	if opts.ScheduleInterval == 0 {
+		it, err := time.ParseDuration(os.Getenv("QUEUE_SCHEDULER_INTERVAL"))
+		if err != nil {
+			it = 1 * time.Second
```
Comment (jeroiraz) [inline adapters/repos/db/queue/scheduler.go:70]: minor: it may be worth at least logging about a potential invalid value set in the env variable?

File: adapters/repos/db/queue/scheduler.go [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+	s.queues.Lock()
+	defer s.queues.Unlock()
+
+	s.queues.m[q.ID()] = newQueueState(q)
+
+	q.Metrics().Registered(q.ID())
+}
+
+func (s *Scheduler) UnregisterQueue(id string) {
+	s.PauseQueue(id)
```
Comment (jeroiraz) [inline adapters/repos/db/queue/scheduler.go:95]: should this be done after the validation of a valid queue id?

File: adapters/repos/db/queue/scheduler.go [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+
+	q := s.getQueue(id)
+	if q == nil {
+		return
+	}
+
+	// wait for the workers to finish processing the queue's tasks
+	q.activeTasks.Wait()
```
Comment (jeroiraz) [inline adapters/repos/db/queue/scheduler.go:103]: still don't know how `activeTasks` are increased but I'm wondering if this steps will be completed if there is a continuous enqueuing

File: adapters/repos/db/queue/scheduler.go [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+		return nil
+	}
+
+	// check if the scheduler is already closed
+	if s.ctx.Err() != nil {
+		return nil
+	}
+
+	// stop scheduling
+	s.cancelFn()
```
Comment (jeroiraz) [inline adapters/repos/db/queue/scheduler.go:161]: just to validate, cancelling the context won't prevent all the active tasks to be completed or if it does it's not a problem to not complete all of them...

File: adapters/repos/db/queue/scheduler.go [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+
+	// wait for the workers to finish processing tasks
+	s.activeTasks.Wait()
+
+	// wait for the spawned goroutines to stop
+	s.wg.Wait()
+
+	// close the channels
+	for _, ch := range s.chans {
+		close(ch)
```
Comment (jeroiraz) [inline adapters/repos/db/queue/scheduler.go:171]: wondering if it's not "responsability" of the workers to close the channels themselves

File: adapters/repos/db/queue/scheduler.go [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			// stop the ticker
+			t.Stop()
+			return
+		case <-t.C:
+			s.schedule()
+		case ch := <-s.triggerCh:
```
Comment (jeroiraz) [inline adapters/repos/db/queue/scheduler.go:241]: triggerCh might be a single channel of the scheduler and sending empty struct{} instance when manually triggering it

File: adapters/repos/db/queue/scheduler.go [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+			return
+		}
+
+		q := s.getQueue(id)
+		if q == nil {
+			continue
+		}
+
+		// mark it as scheduled
+		q.MarkAsScheduled()
```
Comment (jeroiraz) [inline adapters/repos/db/queue/scheduler.go:297]: in this point q is not nil but there is no warranty it's still in the list of queues

File: adapters/repos/db/queue/scheduler.go [hunk 12]
```diff
@@ -0,0 +1,490 @@
...
+		}
+
+		if q.q.Size() == 0 {
+			q.MarkAsUnscheduled()
+			continue
+		}
+
+		count, err := s.dispatchQueue(q)
+		if err != nil {
+			s.Logger.WithError(err).WithField("id", id).Error("failed to schedule queue")
```
Comment (jeroiraz) [inline adapters/repos/db/queue/scheduler.go:319]: what's the impact of this error being received?

File: adapters/repos/db/queue/worker.go [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+	ch := make(chan Batch)
+
+	return &Worker{
+		logger:        logger,
+		retryInterval: retryInterval,
+		ch:            ch,
+	}, ch
+}
+
+func (w *Worker) Run(ctx context.Context) {
```
Comment (jeroiraz) [inline adapters/repos/db/queue/worker.go:38]: Run might return error

File: adapters/repos/db/queue/scheduler.go (hunk 7 above)
Comment (asdine) [reply adapters/repos/db/queue/scheduler.go:103]: Pausing the queue will prevent the scheduler from dequeuing it, not enqueuing. In other terms, it's possible to enqueue while the queue is paused.
`q.activeTasks` only keeps track of the tasks that have already been dispatched to the workers, so calling `Wait` eventually returns event with continuous enqueuing

File: adapters/repos/db/queue/scheduler.go (hunk 8 above)
Comment (asdine) [reply adapters/repos/db/queue/scheduler.go:161]: Canceling the context will cancel all active tasks, as the context is passed to the workers.
This is not a problem because the tasks are supposed to be idempotent

File: adapters/repos/db/queue/scheduler.go [hunk 14]
```diff
@@ -0,0 +1,490 @@
...
+		case <-ctx.Done():
+		}
+	default:
+	}
+}
+
+func (s *Scheduler) schedule() {
+	// as long as there are tasks to schedule, keep running
+	// in a tight loop
+	for {
```
[the remark this answers is not in the document]
Comment (asdine) [reply adapters/repos/db/queue/scheduler.go:264]: the problem is that `runScheduler` uses a ticker, which means that it will only call the scheduling function at most once per second.
This is fine if queues are empty, but if there are not, we want to empty them as fast as possible, without waiting for the ticker to trigger again.

Using two loops is one way to do so: the outer uses the ticker, the inner is tight as long as there are stuff to dequeue

File: adapters/repos/db/queue/scheduler.go (hunk 12 above)
Comment (asdine) [reply adapters/repos/db/queue/scheduler.go:319]: It means the queue couldn't dequeue a batch. If that happens, it means the queue implementation has a problem.
For the DiskQueue implementation, it probably means a disk issue. For future implementations, it could be anything (network, bug, etc.)
In any case, since the scheduler is central, it should not fail. It should simply skip the queue and move on to the next.

File: adapters/repos/db/clusterintegrationtest/cluster_integration_test.go [hunk 15]
```diff
@@ -157,6 +157,15 @@ func testDistributed(t *testing.T, dirName string, rnd *rand.Rand, batch bool) {
 			}
 		})
 	}
+
+	t.Run("wait for indexing to finish", func(t *testing.T) {
+		for _, node := range nodes {
+			time.Sleep(100 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (antas-marcin) [reply adapters/repos/db/clusterintegrationtest/cluster_integration_test.go:163]: Yeah, I second that!

File: adapters/repos/db/index_integration_test.go [hunk 16]
```diff
@@ -688,7 +713,7 @@ func TestIndex_DebugResetVectorIndexTargetVector(t *testing.T) {
 
 func TestIndex_DebugResetVectorIndexPQ(t *testing.T) {
 	t.Setenv("ASYNC_INDEXING", "true")
-	t.Setenv("ASYNC_INDEX_INTERVAL", "100ms")
```
Comment (antas-marcin) [inline adapters/repos/db/index_integration_test.go:691]: should we keep the old name for backward compatability?

File: adapters/repos/db/queue/queue.go [hunk 17]
```diff
@@ -0,0 +1,1063 @@
...
+)
+
+const (
+	// defaultChunkSize is the maximum size of each chunk file. Set to 100MB.
+	defaultChunkSize = 100 * 1024 * 1024
+
+	// chunkFileFmt is the format string for the chunk files,
+	chunkFileFmt = "chunk-%d.bin"
+
+	magicHeader = "WV8Q"
```
Comment (antas-marcin) [inline adapters/repos/db/queue/queue.go:39]: what is a magic header?

File: adapters/repos/db/queue/queue.go [hunk 18]
```diff
@@ -0,0 +1,1063 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	// if there are no more chunks to read,
+	// check if the partial chunk is stale (e.g no tasks were pushed for a while)
+	if c == nil || c.f == nil {
+		c, err = q.checkIfStale()
+		if c == nil || err != nil || c.f == nil {
+			return nil, err
```
Comment (antas-marcin) [inline adapters/repos/db/queue/queue.go:291]: hmm, so here we could potencially have a situation where `c == nil` and `err == nil` then we would return `nil, nil`? is that expected?

File: adapters/repos/db/queue/queue.go [hunk 19]
```diff
@@ -0,0 +1,1063 @@
...
+
+	return nil
+}
+
+func (q *DiskQueue) readChunkRecordCount(path string) (uint64, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return 0, err
+	}
+	defer f.Close()
```
Comment (antas-marcin) [inline adapters/repos/db/queue/queue.go:521]: shouldn't this `defer f.Close()` be put just after the 517 line so that the file handler would be also closed in case of error?

File: adapters/repos/db/queue/queue.go [hunk 20]
```diff
@@ -0,0 +1,1063 @@
...
+	c.r.Reset(c.f)
+	c.size = uint64(stat.Size())
+
+	// check the header
+	c.count, err = readChunkHeader(c.r)
+	if err != nil {
+		return nil, err
+	}
+
+	return &c, nil
```
Comment (antas-marcin) [inline adapters/repos/db/queue/queue.go:582]: we don't need to close `c.f` at the end of this method? we are closing it on line 559, but nowhere else in this method

File: adapters/repos/db/index_integration_test.go (hunk 16 above)
Comment (asdine) [reply adapters/repos/db/index_integration_test.go:691]: I'm not sure as it's not the same logic anymore and also these env vars are not part of the public API. They are primarily used for tests for now.
Once we make async indexing GA we'll communicate on which env var is supported

File: adapters/repos/db/queue/queue.go (hunk 17 above)
Comment (asdine) [reply adapters/repos/db/queue/queue.go:39]: it's a series of 4 bytes that indicate what kind of file this is. It is commonly used to detect the type of the binary file.
https://en.wikipedia.org/wiki/List_of_file_signatures
In this case it's 3 bytes for Weaviate (WV8) and one for the queue (Q). The choice of bytes is completely arbitrary.

File: adapters/repos/db/queue/queue.go (hunk 18 above)
Comment (asdine) [reply adapters/repos/db/queue/queue.go:291]: yeah both case are expected 👍🏼

File: adapters/repos/db/queue/queue.go [hunk 21]
```diff
@@ -0,0 +1,1063 @@
...
+
+		// read the record
+		if cap(buf) < int(length) {
+			buf = make([]byte, length)
+		} else {
+			buf = buf[:length]
+		}
+		_, err = io.ReadFull(c.r, buf)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to read record:")
```
[the remark this answers is not in the document]
Comment (asdine) [reply adapters/repos/db/queue/queue.go:329]: Good catch

File: adapters/repos/db/shard_async.go [hunk 22]
```diff
@@ -50,51 +82,57 @@ func (s *Shard) PreloadQueue(targetVector string) error {
...
+			ID:     id,
+			Vector: vector,
 		}
-
 		counter++
-		return q.Push(ctx, desc)
+
+		batch = append(batch, rec)
+
+		if len(batch) < 1000 {
```
Comment (antas-marcin) [inline adapters/repos/db/shard_async.go:106]: hmm, maybe we could make a const out of this number? tis is just an internal batch size for shard_async? I mean it should not be adjusted in anyways and value 1000 is the right value, yes?

File: adapters/repos/db/queue/queue.go [hunk 23]
```diff
@@ -0,0 +1,1063 @@
...
+
+type BeforeScheduleHook interface {
+	BeforeSchedule() bool
+}
+
+var _ Queue = &DiskQueue{}
+
+type DiskQueue struct {
+	// Logger for the queue. Wrappers of this queue should use this logger.
+	Logger           logrus.FieldLogger
```
[the remark this answers is not in the document]
Comment (asdine) [reply adapters/repos/db/queue/queue.go:60]: yes so it can be used by types that embed the DiskQueue, this is what the `VectorIndexQueue` does for example.
That prevents them from creating their own logger.

File: adapters/repos/db/shard_shutdown.go [hunk 24]
```diff
@@ -97,6 +103,8 @@ func (s *Shard) Shutdown(ctx context.Context) (err error) {
 			}
 		}
 	} else {
+		err = s.queue.Flush()
+		ec.AddWrap(errors.Wrap(err, "flush vector index queue commitlog"))
```
Comment (antas-marcin) [inline adapters/repos/db/shard_shutdown.go:107]: can we here wrap the error without this additional wrap? like so:
```
ec.AddWrap(err, "flush vector index queue commitlog")
```

File: adapters/repos/db/shard_test.go [hunk 25]
```diff
@@ -231,7 +231,7 @@ func TestShard_InvalidVectorBatches(t *testing.T) {
 
 func TestShard_DebugResetVectorIndex(t *testing.T) {
 	t.Setenv("ASYNC_INDEXING", "true")
-	t.Setenv("ASYNC_STALE_TIMEOUT", "200ms")
```
Comment (antas-marcin) [inline adapters/repos/db/shard_test.go:234]: same here, should we keep this env variable for backward compatibility reasons? or it's ok to replace it with a new

File: adapters/repos/db/queue/queue.go (hunk 19 above)
Comment (asdine) [reply adapters/repos/db/queue/queue.go:521]: In general it's better to do it after the err check because most of the time the APIs who return a ressource always return nil in case of error (and they are responsible for cleaning up)

File: adapters/repos/db/queue/queue.go (hunk 20 above)
Comment (asdine) [reply adapters/repos/db/queue/queue.go:582]: We don't need to close it because the role of this function is to open a file and return the file descriptor.
The branch L.556 is in case we are trying to open an empty file (because of a crash for example), it's a self cleaning condition that removes the file and moves on (in that case we need to close it and return nil)

File: adapters/repos/db/shard_async.go (hunk 22 above)
Comment (asdine) [reply adapters/repos/db/shard_async.go:106]: yeah it's just an arbitrary value for this function only, any value not too big and not too small works too. Happy to create a const for it

File: adapters/repos/db/vector_index_queue.go [hunk 26]
```diff
@@ -0,0 +1,379 @@
...
+			op:  op,
+			id:  uint64(id),
+			idx: v.q.vectorIndex,
+		}, nil
+	}
+
+	return nil, errors.Errorf("unknown operation: %d", op)
+}
+
+var _ queue.TaskGrouper = &Task{}
```
Comment (antas-marcin) [inline adapters/repos/db/vector_index_queue.go:301]: same here is this needed?

File: adapters/repos/db/shard_shutdown.go (hunk 24 above)
Comment (asdine) [reply adapters/repos/db/shard_shutdown.go:107]: oh good point

File: adapters/repos/db/shard_status.go [hunk 27]
```diff
@@ -44,7 +43,38 @@ func (s *Shard) GetStatus() storagestate.Status {
...
 
-	return s.status.Status
+	if s.status.Status != storagestate.StatusReady && s.status.Status != storagestate.StatusIndexing {
+		return s.status.Status
+	}
+
+	if s.hasTargetVectors() {
+		if len(s.queues) == 0 {
+			return s.status.Status
+		}
```
[the remark this answers is not in the document]
Comment (asdine) [reply adapters/repos/db/shard_status.go:53]: Could there be a case where we are calling this method during initialization, before we create the queues?

File: test/acceptance/graphql_resolvers/setup_test.go [hunk 28]
```diff
@@ -60,6 +62,8 @@ func testGraphQL(t *testing.T) {
 	// tests with classes that have objects with same uuids
 	t.Run("import test data (near object search class)", addTestDataNearObjectSearch)
 
+	waitForIndexing()
```
Comment (antas-marcin) [inline test/acceptance/graphql_resolvers/setup_test.go:65]: can we change it to `assertEventuallyWIth` like I did this [here](https://github.com/weaviate/weaviate/blob/main/test/acceptance/graphql_resolvers/setup_test.go#L1108-L1124)? so just to avoid sleeps which tend to convert to flaky tests

Comment (asdine) [reply test/acceptance/graphql_resolvers/setup_test.go:65]: yes great point

File: adapters/repos/db/queue/queue.go [hunk 29]
```diff
@@ -0,0 +1,1063 @@
...
+	Size() int64
+	DequeueBatch() (batch *Batch, err error)
+	Metrics() *Metrics
+}
+
+type BeforeScheduleHook interface {
+	BeforeSchedule() bool
+}
+
+var _ Queue = &DiskQueue{}
```
[the remark this answers is not in the document]
Comment (antas-marcin) [reply adapters/repos/db/queue/queue.go:56]:
```
var _ = Queue(&DiskQueue{})
```

File: adapters/repos/db/queue/scheduler.go (hunk 9 above)
Comment (asdine) [reply adapters/repos/db/queue/scheduler.go:194]: Had the same idea but it's safer to let the writer close to avoid panics

File: adapters/repos/db/queue/scheduler.go (hunk 10 above)
Comment (asdine) [reply adapters/repos/db/queue/scheduler.go:284]: The issue is that we need two way communication, one way to trigger, one way to notify that the job is done

File: adapters/repos/db/queue/scheduler.go [hunk 30]
```diff
@@ -0,0 +1,490 @@
...
+	for id := range s.queues.m {
+		ids = append(ids, id)
+	}
+	s.queues.Unlock()
+
+	nothingScheduled = true
+
+	for _, id := range ids {
+		if s.ctx.Err() != nil {
+			return
```
[the remark this answers is not in the document]
Comment (asdine) [reply adapters/repos/db/queue/scheduler.go:336]: Even if it's returned it will be ignored as it's not technically an error

File: adapters/repos/db/queue/scheduler.go (hunk 11 above)
Comment (asdine) [reply adapters/repos/db/queue/scheduler.go:345]: Exactly, and it's totally fine as this can also happen lower in the chain

File: adapters/repos/db/queue/worker.go (hunk 13 above)
Comment (asdine) [reply adapters/repos/db/queue/worker.go:38]: It doesn't need to actually, it's a very simple function and it's spawned in a goroutine

**Review by trengrj — APPROVED** [review]
